MotionMark/resources/runner/results.js

```diff
@@ -0,0 +1,468 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+class RunData {
+    constructor(version, options, runs = [])
+    {
+        this.version = version;
+        this.options = options;
+        this.runs = runs;
+    }
+    
+    static resultsDataFromSingleRunData(singleRunData)
+    {
+        RunData.#migrateImportedData(singleRunData);
+        
+        if (!singleRunData.data instanceof Array) {
+            console.log('Imported singleRunData.data is not an array. Bailing');
+            return null;
+        }
+        
+        return new RunData(singleRunData.version, singleRunData.options, singleRunData.data);
+    }
+
+    static resultsDataFromBenchmarkRunnerData(benchmarkData)
+    {
+        if (!benchmarkData instanceof Array) {
+            console.log('Imported benchmarkData is not an array. Bailing');
+            return null;
+        }
+        
+        let runData = [];
+        for (let run of benchmarkData) {
+            RunData.#migrateImportedData(run);
+            if (run.data.length !== 1) {
+                console.log('Imported benchmarkData has a "data" array with an unexpected number of items. Bailing');
+                return null;
+            }
+        
+            runData.push(run.data[0]);
+        }
+        
+        // Version and options data should be these same for each run. Use the first run's information.
+        return new RunData(benchmarkData[0].version, benchmarkData[0].options, runData);
+    }
+    
+    static #migrateImportedData(options)
+    {
+        if (!("version" in options))
+            options.version = "1.0";
+        
+        if (!("frame-rate" in options)) {
+            options.options["frame-rate"] = 60;
+            console.log("No frame-rate data; assuming 60fps")
+        }
+
+        if (!("system-frame-rate" in options)) {
+            options.options["system-frame-rate"] = 60;
+            console.log("No system-frame-rate data; assuming 60fps")
+        }
+    }
+}
+
+class ScoreCalculator {
+    constructor(runData)
+    {
+        this._runData = runData;
```

Comment:
Why not use private member variables?

Reply:
The whole codebase needs a switch to private variables. For now I was following existing code.